src/main/java/frc/robot/subsystems/SwerveModule.java

```diff
@@ -0,0 +1,234 @@
+import static edu.wpi.first.units.Units.Volts;
+
+public class SwerveModule {
+
+
+    private static final double DRIVE_GEAR_RATIO = 6.75;
+    private static final double STEER_GEAR_RATIO = 12.8;
+    private static final double WHEEL_RADIUS = 0.0508;
+    private static final double WHEEL_CIRCUMFERENCE = 2*Math.PI*WHEEL_RADIUS;
+
+    private static final double STEER_P = 0.56; // 0.05 //0.052 0053 0.054
+    private static final double STEER_I = 0.00003; //0
+    private static final double STEER_D = 0.01;
+    private static final double STEER_F = 0.;
+
+    private static final double DRIVE_P = 0.0001;//0.0003;
+    private static final double DRIVE_I = 0;//5e-7;
+    private static final double DRIVE_D = 0;//5e-7;
+    private static final double DRIVE_F = 0.00016;//0.0001;
+
+    private final CANSparkMax drive;
+    private final CANSparkMax steer;
+    private final CANcoder canCoder;
+
+    private final RelativeEncoder driveEncoder;
+    private final RelativeEncoder steerEncoder;
+    private final SparkPIDController drivePID;
+    private final SparkPIDController steerPID;
+
+    private final String printName;
+
+    public SwerveModule(CANSparkMax drive,
+                        CANSparkMax steer,
+                        CANcoder canCoder,
+                        double zeroAngle,
+                        String identifier){
+        this.drive = drive;
+        this.steer = steer;
+        this.canCoder = canCoder;
+
+        this.printName = String.format("SwerveModule[%s]", identifier);
+
+        this.drive.restoreFactoryDefaults();
+        this.steer.restoreFactoryDefaults();
+
+        this.driveEncoder = this.drive.getEncoder();
+        this.steerEncoder = this.steer.getEncoder();
+        this.drivePID = this.drive.getPIDController();
+        this.steerPID = this.steer.getPIDController();
+
+        driveEncoder.setPosition(0);
+        steerEncoder.setPosition(0);
+
+        drivePID.setP(DRIVE_P);
+        drivePID.setI(DRIVE_I);
+        drivePID.setD(DRIVE_D);
+        drivePID.setFF(DRIVE_F);
+        drivePID.setOutputRange(-1, 1);
+
+        steerPID.setP(STEER_P);
+        steerPID.setI(STEER_I);
+        steerPID.setD(STEER_D);
+        steerPID.setFF(STEER_F);
+        steerPID.setOutputRange(-1, 1);
+        steerPID.setPositionPIDWrappingMaxInput(STEER_GEAR_RATIO);
+        steerPID.setPositionPIDWrappingMinInput(0);
+        steerPID.setPositionPIDWrappingEnabled(true);
+
+        steerEncoder.setPosition((this.canCoder.getAbsolutePosition().refresh().getValue() - zeroAngle) / 360 * STEER_GEAR_RATIO);
+    }
+
+    public double getHeadingDegrees() {
+        double degrees = steerEncoder.getPosition() / STEER_GEAR_RATIO * 360;
+        degrees %= 360;
+        if (degrees < 0) {
+            degrees += 360;
+        }
+
+        return degrees;
+    }
+
+    private double getAbsoluteHeadingDegrees() {
+        return canCoder.getAbsolutePosition().refresh().getValue();
```

Comment:
you should store the `StatusSignal` returned from `canCoder.getAbsolutePosition()` in the ctor and use that stored variable from then.

This is because the initial retrieval of a status signal has some overhead it would seem. So calling `getAbsolutePosition` repeatably would have some impact (although small).

Reply:
didnt raelly understood that 100%, ill text u

Reply:
i used statusSignal = canCoder.getAbsolutePosition();in the ctor
then what do i do in the function ?

Reply:
```suggestion
        return statusSignal.refresh().getValue();
```